### PR TITLE
CB-13498 Consume new CM api to bulk decommission hosts (hosts/removeH…

### PR DESCRIPTION
**Pull request by lturcsanyi**
…ostsFromCluster).

The new API is available from 7.2.13 (CM version 7.5.3), before that it will use the old approach.

See detailed description in the commit message.

### CODE REVIEW

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/DecommissionHandler.java [hunk 1]
```diff
@@ -144,6 +153,45 @@ public void accept(Event<DecommissionRequest> event) {
...
 
+    private DecommissionResult bulkHostsRemoval(DecommissionRequest request, Set<String> hostNames, boolean forced, Stack stack,
+            ClusterDecomissionService clusterDecomissionService, Map<String, InstanceMetaData> hostsToRemove)
+            throws CloudbreakOrchestratorFailedException, CloudbreakException {
+        try {
+            clusterDecomissionService.removeHostsFromCluster(Lists.newArrayList(hostsToRemove.values()));
+        } catch (CloudbreakException e) {
+            if (e.getMessage() != null && e.getMessage().contains("bulk host removal is not supported")) {
+                LOGGER.warn("Bulk host removal was unsuccessful, fallback to single host removal.");
+                return singleHostsRemoval(request, hostNames, forced, stack, clusterDecomissionService, hostsToRemove);
```
Comment (horadla23) [inline core/src/main/java/com/sequenceiq/cloudbreak/reactor/DecommissionHandler.java:164]: may worth to throw a specific exception for this case and catch it within different catch block

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerDecomissioner.java [hunk 2]
```diff
@@ -328,6 +336,35 @@ public void deleteUnusedCredentialsFromCluster(Stack stack, ApiClient client) {
...
+            List<String> knownDeletableHosts = hostRefList.getItems().stream()
+                    .filter(host -> hosts.contains(host.getHostname()))
+                    .map(ApiHost::getHostname)
+                    .collect(Collectors.toList());
+            if (!knownDeletableHosts.isEmpty()) {
+                ApiHostsToRemoveArgs body = new ApiHostsToRemoveArgs();
+                body.hostsToRemove(knownDeletableHosts);
+                body.deleteHosts(Boolean.TRUE);
+                ApiCommand command = hostsResourceApi.removeHostsFromCluster(body);
+                LOGGER.debug("Hosts remove request sent.");
```
Comment (horadla23) [inline cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerDecomissioner.java:352]: I would extend this log message with some information (like hosts list)

**Review by horadla23 — APPROVED** [review]
I had just one minor comment regarding logging

**Comment by lajosrodek** [comment]
Achievement unlocked: Commit over 100K changed lines. :smile:

**Review by topolyai5 — CHANGES_REQUESTED** [review]

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterDecommissionService.java [hunk 3]
```diff
@@ -63,53 +66,68 @@ public void initApiClient() throws ClusterClientInitException {
...
         try {
-            client = clouderaManagerApiClientProvider.getV31Client(stack.getGatewayPort(), user, password, clientConfig);
+            v31Client = clouderaManagerApiClientProvider.getV31Client(stack.getGatewayPort(), user, password, clientConfig);
         } catch (ClouderaManagerClientInitException e) {
             throw new ClusterClientInitException(e);
         }
+        try {
+            v45Client = clouderaManagerApiClientProvider.getV45Client(stack.getGatewayPort(), user, password, clientConfig);
+        } catch (ClouderaManagerClientInitException e) {
+            LOGGER.error("Client init failed for V45 client, thus bulk host removal is not supported");
```
[the remark this answers is not in the document]
Comment (lturcsanyi) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterDecommissionService.java:76]: I would like to see if something went wrong and CB could not init v45 client, thus I would stick with error or at least warn log level.

[the remark this answers is not in the document]
Comment (topolyai5) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterDecommissionService.java:76]: If you don't throw an error it must be maximum a warn. But it does not affect the running and you throw the exception during the bulk hist method. It is enough debug level for this message

[the remark this answers is not in the document]
Comment (lturcsanyi) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterDecommissionService.java:76]: If we cannot init the client it means something definitely went wrong, I'll change this to warn.

**Review by horadla23 — APPROVED** [review]

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterDecommissionService.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (topolyai5) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerClusterDecommissionService.java]: ok, warn will be good

**Review by topolyai5 — APPROVED** [review]

**Pull request merged**